runtime/CMakeLists.txt

```diff
@@ -49,6 +49,7 @@ endif()
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/include)
 
 add_subdirectory(ompstub)
+add_subdirectory(libpgmath)
```

Comment:
I don't know how to accept this change & build that with different compilers for flang & libpgmath.

Reply:
If building libpgmath with a different compiler than the rest of flang is a requirement, a reasonable way of doing this would be:

1. Allow (but not require) the definition of libpgmath-specific compilers in CMAKE_OPTIONS:

-DLIBPGMATH_C_COMPILER:FILEPATH=<insert-c-compiler-path-here>
-DLIBPGMATH_CXX_COMPILER:FILEPATH=<insert-c++-compiler-path-here>
-DLIBPGMATH_CFLAGS:STRING=<insert-libpgmath-c-flags-here>
-DLIBPGMATH_CXXFLAGS:STRING=<insert-libpgmath-c++-flags-here>

2. Add the appropriate logic in libpgmath's toplevel CMakeLists.txt file:

if ("${LIBPGMATH_C_COMPILER}" STREQUAL "")
else()
  set(${CMAKE_C_COMPILER} "${LIBPGMATH_C_COMPILER}")
endif()

if ("${LIBPGMATH_CXX_COMPILER}" STREQUAL "")
else()
  set(${CMAKE_CXX_COMPILER} "${LIBPGMATH_CXX_COMPILER}")
endif()

if ("${LIBPGMATH_CFLAGS}" STREQUAL "")
else()
  set(${CMAKE_C_FLAGS} "${LIBPGMATH_CFLAGS}")
endif()

if ("${LIBPGMATH_CXXFLAGS}" STREQUAL "")
else()
  set(${CMAKE_CXX_FLAGS} "${LIBPGMATH_CXXFLAGS}")
endif()

[ ... etc etc etc ... ]

3. I am skeptical that LLVM will be open to the concept of building flang with one compiler and building libpgmath with another. That's not something they have ever allowed for LLVM or clang.

Reply:
Also:

> %> pwd
> ${top_srcdir}/tools/flang/runtime

> %> diff -u CMakeLists.txt.orig CMakeLists.txt
> --- CMakeLists.txt.orig 2018-06-20 14:47:29.239660258 +0000
> +++ CMakeLists.txt      2018-06-21 18:24:09.446198990 +0000
> @@ -49,7 +49,6 @@
 > include_directories(${CMAKE_CURRENT_SOURCE_DIR}/include)
 >
 > add_subdirectory(ompstub)
> -add_subdirectory(libpgmath)
 >
 > add_subdirectory(flangrti)
 > add_subdirectory(flang)

*Without* the explicit

> add_subdirectory(libpgmath)

line in ${top_srcdir}/tools/flang/runtime/CMakeLists.txt, the following happens:

> cd /builds-1/steleman/llvm-git/20180620-flang-github-cavium-60/build/tools/flang/runtime/flang && /usr/local/gcc73/bin/gcc  -DLINUX -DMAXCPUS=256 -DMAXCPUSL=8 -DMAXCPUSR=8 -DNATIVE_FPCVT -DPGF90 -DPGFLANG -DPGI_LITTLE_ENDIAN -DTARGET_LINUX -DTARGET_LINUX_ARM -DTARGET_LLVM -DTARGET_LLVM_ARM64 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Dflang_shared_EXPORTS -DDESC_I8 -DHAVE_LONG_LONG_INT -DINT32PTR64 -DTM_I8 -I/builds-1/steleman/llvm-git/20180620-flang-github-cavium-60/build/tools/flang/runtime/flang -I/builds-1/steleman/llvm-git/20180620-flang-github-cavium-60/llvm-git/tools/flang/runtime/flang -I/builds-1/steleman/llvm-git/20180620-flang-github-cavium-60/llvm-git/tools/flang/include -I/builds-1/steleman/llvm-git/20180620-flang-github-cavium-60/build/tools/flang/include -I/usr/include/libxml2 -I/builds-1/steleman/llvm-git/20180620-flang-github-cavium-60/build/include -I/builds-1/steleman/llvm-git/20180620-flang-github-cavium-60/llvm-git/include -I/builds-1/steleman/llvm-git/20180620-flang-github-cavium-60/llvm-git/tools/flang/runtime/include  -g -O2 -pthread -fno-strict-aliasing -fno-tree-slp-vectorize -fno-tree-vectorize -fno-reorder-blocks -fno-toplevel-reorder -Wall -Wcast-align -Wno-long-long -Woverflow -Wstack-protector -ffunction-sections -fdata-sections -fkeep-static-consts -fstack-protector-all -fno-omit-frame-pointer -Wl,-rpath -Wl,/builds-1/steleman/llvm-git/20180620-flang-github-cavium-60/build/lib -Wl,-rpath -Wl,/usr/local/gcc73/lib64  -fPIC -Werror=date-time -Wall -W -Wno-unused-parameter -Wwrite-strings -Wno-missing-field-initializers -pedantic -Wno-long-long -Wno-comment -ffunction-sections -fdata-sections -O2 -DNDEBUG -fPIC   -fPIC -mcpu=thunderx2t99 -o CMakeFiles/flang_shared.dir/I8_sources/type.c.o   -c /builds-1/steleman/llvm-git/20180620-flang-github-cavium-60/build/tools/flang/runtime/flang/I8_sources/type.c
> gmake[2]: *** No rule to make target 'lib/libpgmath.so', needed by 'lib/libflang.so'.  > Stop.
> gmake[2]: *** Waiting for unfinished jobs....

Reply:
We'd like to keep the libpgmath build separate from the rest of the flang build.  The intention is to promote libpgmath to the top-level of flang-compiler.  The readme and other docs describe the separate build-step for libpgmath.

Reply:
OK, but I don't see how you can do that and build flang at the same time.

As shown in my copy-and-paste example above, without the explicit

> add_subdirectory(libpgmath)

line, flang won't build.